cartographer/sensor/rangefinder_point.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_SENSOR_RANGEFINDER_POINT_H_
+#define CARTOGRAPHER_SENSOR_RANGEFINDER_POINT_H_
+
+#include <vector>
+
+#include "Eigen/Core"
+#include "cartographer/sensor/proto/sensor.pb.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace sensor {
+
+// Stores 3D position of a point observed by a rangefinder sensor.
```

Comment:
1. Please, use composition here. 
2. All data members should be in `private` section of the class. 
3. Try to get rid of friend functions.

Reply:
About the friend `operator*`: we typically apply a multiplication with a `Rigid3` pose estimate from the left. Unfortunately, a member `operator*` only allows `this` as the left operand. So I see two solutions: either an overloaded friend `operator*` like this, or a member function e.g. `RangefinderPoint::ToLocal(Rigid3 pose_estimate)` which would do a left multiplication.

Reply:
Also, about composition: you mentioned you would inline the `Vector3f` in `TimedRangefinderPoint` proto. Does this mean that the C++ class should be the same?

Slicing a `TimedRangefinderPoint` into a `RangefinderPoint` then means a little bit of more work. Should I put it in a `TimedRangefinderPoint::ToRangefinderPoint` member?

Reply:
Make them structs. As in proto. 

```
struct TimedRangefinderPoint {
  proto::RangefinderPoint ToProto() const;
  Eigen::Vector3f position_;
  float time_;
};

struct RangefinderPoint {
  proto::RangefinderPoint ToProto() const;
  Eigen::Vector3f position_;
};
```

Reply:
Isn't `_` only for private members? Also, I think structs shouldn't have methods, IIRC `ToProto/FromProto` are implemented as functions outside structs.

Reply:
In structs, you do not need `_`, sorry, i just copied. Structs should not have complicated logic, yes. But serialization/deserialization can be put there. Up to you.

Reply:
Okay, cool. I'm only missing one more detail: where to implement discarding time? Member `TimedRangefinderPoint::ToRangefinderPoint` or standalone function?